File: src/p0002/python/solution.py
class Solution:
    def addTwoNumbers(self, a: ListNode, b: ListNode) -> ListNode:
        sentinel = ListNode(0)
        tail = sentinel
        carry = 0
        while carry or a or b:
            if a:
                carry, a = carry + a.val, a.next
            if b:
                carry, b = carry + b.val, b.next
            tail.next = ListNode(carry % 10)
            tail = tail.next
            carry = carry // 10
        return sentinel.next

File: src/p0002/rust/solution.rs
use std::ops::DerefMut;
use std::borrow::BorrowMut;

impl Solution {
    pub fn add_two_numbers(p: Option<Box<ListNode>>, q: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let (mut a, mut b, mut carry) = (p, q, 0);
        let mut sentinel = ListNode::new(0);
        let mut tail = sentinel.borrow_mut();
        while carry != 0 || a.is_some() || b.is_some() {
            if let Some(t) = a { carry += t.val; a = t.next; }
            if let Some(t) = b { carry += t.val; b = t.next; }
            tail.next = Some(Box::new(ListNode::new(carry % 10)));
            tail = tail.next.as_mut().unwrap().deref_mut();
            carry /= 10;
        }
        sentinel.next
    }
}
